bpfman-operator/apis/v1alpha1/cluster_bpf_application_types.go

```diff
@@ -0,0 +1,166 @@
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Kprobe' ?  has(self.kprobeInfo) : !has(self.kprobeInfo)",message="kprobeInfo configuration is required when type is kprobe, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'KretProbe' ?  has(self.kretprobeInfo) : !has(self.kretprobeInfo)",message="kretprobeInfo configuration is required when type is kretprobe, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Uprobe' ?  has(self.uprobeInfo) : !has(self.uprobeInfo)",message="uprobeInfo configuration is required when type is uprobe, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'UretProbe' ?  has(self.uretprobeInfo) : !has(self.uretprobeInfo)",message="uretprobeInfo configuration is required when type is uretprobe, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'TracePoint' ?  has(self.tracepointInfo) : !has(self.tracepointInfo)",message="tracepointInfo configuration is required when type is tracepoint, and forbidden otherwise"
+type ClBpfApplicationProgram struct {
+	// name is the name of the function that is the entry point for the BPF
+	// program
+	Name string `json:"name"`
+
+	// type specifies the bpf program type
+	// +unionDiscriminator
+	// +required
+	// +kubebuilder:validation:Enum:="XDP";"TC";"TCX";"Fentry";"Fexit";"Kprobe";"KretProbe";"Uprobe";"UretProbe";"TracePoint"
+	Type EBPFProgType `json:"type"`
+
+	// xdpInfo defines the desired state of the application's XdpPrograms.
+	// +unionMember
+	// +optional
+	XDPInfo *ClXdpProgramInfo `json:"xdpInfo,omitempty"`
+
+	// tcInfo defines the desired state of the application's TcPrograms.
+	// +unionMember
+	// +optional
+	TCInfo *ClTcProgramInfo `json:"tcInfo,omitempty"`
+
+	// tcxInfo defines the desired state of the application's TcxPrograms.
+	// +unionMember
+	// +optional
+	TCXInfo *ClTcxProgramInfo `json:"tcxInfo,omitempty"`
+
+	// fentryInfo defines the desired state of the application's FentryPrograms.
+	// +unionMember
+	// +optional
+	FentryInfo *ClFentryProgramInfo `json:"fentryInfo,omitempty"`
+
+	// fexitInfo defines the desired state of the application's FexitPrograms.
+	// +unionMember
+	// +optional
+	FexitInfo *ClFexitProgramInfo `json:"fexitInfo,omitempty"`
+
+	// kprobeInfo defines the desired state of the application's KprobePrograms.
+	// +unionMember
+	// +optional
+	KprobeInfo *ClKprobeProgramInfo `json:"kprobeInfo,omitempty"`
+
+	// kretprobeInfo defines the desired state of the application's KretprobePrograms.
+	// +unionMember
+	// +optional
+	KretprobeInfo *ClKprobeProgramInfo `json:"kretprobeInfo,omitempty"`
```

Comment:
It looked like `ClKprobeProgramInfo` had some restrictions for `KretProbe` program types that may not apply to the `Kprobe` program type? Maybe this should use a separate type so that you can have distinct requirements for each type?

Reply:
u mean `offset` value of `0` in case of kretprobe ? we were thinking of just using validation wh to ensure this or maybe we can do that with cel rule but other than that its the same?

Reply:
>u mean offset value of 0 in case of kretprobe ?

Yes

>we were thinking of just using validation wh to ensure this or maybe we can do that with cel rule but other than that its the same?

For now. Is it possible that the `kretprobe` case ends up requiring additional drift from `kprobe` in the future? If you go with using the same type, every change you make for one program type will inherently be added to the other. It is _usually_ better practice to have each one have their own distinct type so you can update each one individually as needed.

Reply:
create kretprobe type